content_scripts/vimium_frontend.coffee

```diff
@@ -133,6 +133,28 @@ class GrabBackFocus extends Mode
     element.blur()
     @suppressEvent
 
+# Pages can load new content dynamically and change the displayed URL using history.pushState. Since this can
+# often be indistinguishable from an actual new page load for the user, we should also re-start GrabBackFocus
+# for these as well.
+handlerStack.push
+  click: (event) ->
+    target = event.target
+    while target
+      # Often, a link which triggers a content load and url change with javascript will also have the new
+      # url as it's href attribute.
+      if target.tagName == "A" and
```

Comment:
> Elsewhere, we use `toLowerCase` for comparisons like this.

For the built-in elements, `tagName` is always upper case, so there's no need to do this.

I've pushed commits dealing with all your other comments.